OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs

```diff
@@ -0,0 +1,205 @@
+						+ "Try using different palettes for resource types that use different blend modes.");
+			}
+
+			foreach (var cell in w.Map.AllCells)
+			{
+				var type = ResourceLayer.GetResourceType(cell);
+				if (type != null && Info.RenderTypes.Contains(type.Info.Type))
+				{
+					// Initialize the RenderContent with the initial map state
+					// because the shroud may not be enabled.
+					var cellContent = CreateResourceCell(type, cell);
+					RenderContent[cell] = cellContent;
+					UpdateRenderedSprite(cell);
+				}
+			}
+		}
+
+		void UpdateSpriteLayers(CPos cell)
+		{
+			var type = ResourceLayer.GetResourceType(cell);
+			if (type != null && !Info.RenderTypes.Contains(type.Info.Type))
+				return;
+
+			var resource = RenderContent[cell];
+			foreach (var kv in spriteLayers)
+			{
+				// resource.Type is meaningless (and may be null) if resource.Sprite is null
+				if (resource.Sprite != null && type.Palette == kv.Key)
+					kv.Value.Update(cell, resource.Sprite);
+				else
+					kv.Value.Update(cell, null);
+			}
+		}
+
+		void IRenderOverlay.Render(WorldRenderer wr)
+		{
+			foreach (var kv in spriteLayers.Values)
+				kv.Draw(wr.Viewport);
+		}
+
+		void ITickRender.TickRender(WorldRenderer wr, Actor self)
+		{
+			var remove = new List<CPos>();
+			foreach (var c in dirty)
+			{
+				if (!self.World.FogObscures(c))
+				{
+					var t = ResourceLayer.GetResourceType(c);
+					if (t != null)
+					{
+						var cellContent = CreateResourceCell(t, c);
+						RenderContent[c] = cellContent;
+					}
+
+					UpdateRenderedSprite(c);
+					remove.Add(c);
+				}
+			}
+
+			foreach (var r in remove)
+				dirty.Remove(r);
+		}
+
+		protected virtual void UpdateRenderedSprite(CPos cell)
+		{
+			var content = ResourceLayer.GetResourceContent(cell);
+			var density = content.Density;
+			var c = RenderContent[cell];
+			if (density > 0)
+			{
+				var type = content.Type;
+				if (type != null && !Info.RenderTypes.Contains(type.Info.Type))
+					return;
+				var sprites = type.Variants[c.Variant];
+				var maxDensity = ResourceLayer.GetMaxResourceDensity(cell);
+				var frame = int2.Lerp(0, sprites.Length - 1, density, maxDensity);
+				c.Sprite = sprites[frame];
+				c.Type = type;
+			}
+			else
+				c.Sprite = null;
+
+			RenderContent[cell] = c;
+		}
+
+		bool disposed;
+		void INotifyActorDisposing.Disposing(Actor self)
+		{
+			if (disposed)
+				return;
+
+			foreach (var kv in spriteLayers.Values)
+				kv.Dispose();
+
+			RenderContent.CellEntryChanged -= UpdateSpriteLayers;
+
+			disposed = true;
+		}
+
+		CellContents CreateResourceCell(ResourceType t, CPos cell)
+		{
+			return new CellContents
+			{
+				Type = t,
+				Variant = ChooseRandomVariant(t)
+			};
+		}
+
+		protected virtual string ChooseRandomVariant(ResourceType t)
+		{
+			return t.Variants.Keys.Random(Game.CosmeticRandom);
+		}
+
+		public ResourceType GetRenderedResourceType(CPos cell) { return RenderContent[cell].Type; }
+
+		public struct CellContents
```

Comment:
Maybe rename to `CellRenderContents` to avoid any possible confusion when reading the code.